File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        backtrack(candidates,0,target,0,new ArrayList<>(), res);
        return res;
    }
    void backtrack(int[] nums, int i, int target, int sum, 
                        List<Integer> p, List<List<Integer>> res){
        if(sum == target){
            for(List l : res){
                if(l.equals(p)){
                    return;
                }
            }
            res.add(new ArrayList<>(p));
            return;
        }
        
        for(int j = i; j<nums.length; j++){
            if(sum + nums[j] > target) break;
            p.add(nums[j]);
            backtrack(nums,j+1,target,sum+nums[j],p,res);
            p.remove(p.size()-1);
        }
        return;
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-2.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums) % 2:
            return False
        n = len(nums)
        target = sum(nums) // 2
        dp = [[-1] * (target+1) for _ in range(n+1)]

        def dfs(i, target):
            if target == 0:
                return True
            if i >= n or target < 0:
                return False
            if dp[i][target] != -1:
                return dp[i][target]

            dp[i][target] = dfs(i+1, target - nums[i]) or dfs(i+1, target)

            return dp[i][target]

        return dfs(0 ,target)